src/compute/src/extensions/temporal_bucket.rs

```diff
@@ -56,46 +66,68 @@ where
+
             // Cap tracking the lower bound of potentially outstanding data.
             let mut cap = Some(cap);
+
+            // Buffer for data to be inserted into the chain.
             let mut buffer = Vec::new();
+
             move |input, output| {
-                while let Some((_time, data)) = input.next() {
+                // The upper frontier is the meet of the input frontier and the `as_of` frontier,
+                // with the `threshold` summary applied to it.
+                let mut upper = Antichain::new();
+                for time1 in &input.frontier().frontier() {
+                    for time2 in as_of.elements() {
+                        // TODO: Use `join_assign` if we ever use a timestamp with allocations.
+                        if let Some(time) = threshold.results_in(&time1.join(time2)) {
+                            upper.insert(time);
+                        }
+                    }
+                }
+
+                while let Some((time, data)) = input.next() {
+                    // Skip data that is about to be revealed.
+                    let pass_through = data.extract_if(.., |(_, t, _)| !upper.less_equal(t));
+                    output
+                        .session_with_builder(&time)
+                        .give_iterator(pass_through);
+
                     // Sort data by time, then drain it into a buffer that contains data for a
-                    // single bucket.
+                    // single bucket. We scan the data for ranges of time that fall into the same
+                    // bucket so we can push batches of data at once.
                     data.sort_by(|(_, t, _), (_, t2, _)| t.cmp(t2));
 
-                    let mut range = None;
-
-                    for (datum, time, diff) in data.drain(..) {
-                        // If we have a range, check if the time is not within it.
-                        if let Some((start, end)) = &range
-                            && (time < *start || time >= *end)
-                        {
-                            // If the time is outside the range, push the current buffer
-                            // to the chain and reset the range.
-                            if !buffer.is_empty() {
-                                let wrapper = chain.find_mut(start).expect("Must exist");
-                                wrapper.inner.push_container(&mut buffer);
-                                buffer.clear();
+                    let mut drain = data.drain(..);
+                    if let Some((datum, time, diff)) = drain.next() {
+                        let mut range = chain.range_of(&time).expect("Must exist");
+                        buffer.push((datum, time, diff));
+                        for (datum, time, diff) in drain {
+                            // If we have a range, check if the time is not within it.
+                            if time < range.0 || time >= range.1 {
```

Comment:
A nice thing you can do for the `range_of` API is to return an actual Range object (it's in std) and here do `range.contains(&time)`